### ISSUE TITLE
Connection timeout

### ISSUE THREAD

**Issue by prophe05** [GideonLeGrange/mikrotik-java#16]
Can you provide possibility to manage connection timout?

**Comment by ruffboy** [comment]
i also have problem with timeout. I have string array with ip's (800) in it.  Some of them are not routeros machines and when loop get to non ros machine code stops on "ApiConnection.connect(ip, port)" ... without any exception or i'm doing something wrong ;)

**Comment by GideonLeGrange** [comment]
@ruffboy Logging on to something that isn't listening on port 8728 should cause a ApiConnectionException. Here is an example of what happens if I try to log on to a Linux server with the API:

```Exception in thread "main" me.legrange.mikrotik.ApiConnectionException: Error connecting to 192.168.1.1:8728 : Connection refused
	at me.legrange.mikrotik.impl.ApiConnectionImpl.open(ApiConnectionImpl.java:190)
	at me.legrange.mikrotik.impl.ApiConnectionImpl.connect(ApiConnectionImpl.java:43)
	at me.legrange.mikrotik.ApiConnection.connect(ApiConnection.java:46)
	at me.legrange.mikrotik.ApiConnection.connect(ApiConnection.java:56)
	at examples.Example.connect(Example.java:12)
	at examples.Example9.main(Example9.java:16)
Caused by: java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:345)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at java.net.Socket.<init>(Socket.java:434)
	at java.net.Socket.<init>(Socket.java:244)
	at me.legrange.mikrotik.impl.ApiConnectionImpl.open(ApiConnectionImpl.java:174)
	... 5 more
```
So I don't understand how you can have a timeout problem on non-RouterOS machines. Can  you please elaborate?

Thanks

**Comment by ruffboy** [comment]
@GideonLeGrange, we have in our network device restarters (circut braker with webinterface, when ROS machine crash, we can restart it with power up/down through web interface) when my loop get's ip of that device it's stops on ApiConnection.connect(ip, port), whitout Connection refused .

        for (String ip : listaIp) {
            try{
                System.out.println("try"+ ip);
                ApiConnection con; 
                con = ApiConnection.connect(ip, 8293); 
                System.out.println(">>: "+ ip);   
                con.login(USER, PASS);
                List<Map<String, String>> rs;
                rs = con.execute(SOME COMMAND);
                con.disconnect();
                System.out.println(rs);
            } catch (MikrotikApiException | InterruptedException e){
                System.out.println(e.getMessage());
            }
        }

**Comment by GideonLeGrange** [comment]
@ruffboy It's still not clear to me why you have a problem. If I try to connect on a port that will never respond (I tested this by adding a "DROP" firewall rule to a Linux server), I eventually get an exception:

```java
Exception in thread "main" me.legrange.mikrotik.ApiConnectionException: Error connecting to 192.168.1.1:8728 : Operation timed out
	at me.legrange.mikrotik.impl.ApiConnectionImpl.open(ApiConnectionImpl.java:190)
	at me.legrange.mikrotik.impl.ApiConnectionImpl.connect(ApiConnectionImpl.java:43)
	at me.legrange.mikrotik.ApiConnection.connect(ApiConnection.java:46)
	at me.legrange.mikrotik.ApiConnection.connect(ApiConnection.java:56)
	at examples.Example.connect(Example.java:12)
	at examples.Example9.main(Example9.java:16)
```

So is the problem that the loop stops for too long, or does it stop for forever? The latter is really unlikely.

**Comment by ruffboy** [comment]
i run the code again and i leave it, you were right  the loop stops for too long time, after 120 000 ms i get exception Connection Timeout, where ROS machine throw this exception in 30 ms :)

**Comment by GideonLeGrange** [comment]
@ruffboy I've implemented connection timeout option which you can try. You will need to check out the bug#16 branch from GitHub and compile that in order to test. If the solution works, I will pull it into the master branch. 

To set the connection timeout, you need to call ```ApiConnection.connect()``` with an extra parameter for the timeout, like this:

```java
    con = ApiConnection.connect(ip, 8293, 3000); // give up after 3 seconds
```

Since you already seem to use a custom port, you can just add the number of milliseconds for the timeout as third parameter. 

Please try it and let me know.

**Comment by ruffboy** [comment]
it works ! :tada:  :)

**Comment by GideonLeGrange** [comment]
@ruffboy Thanks for the feedback. 

@prophe05 The connection timeout I've added could solve some of your issues. I'd appreciate it if you tried the bug#16 branch and let me know. I'm busy implementing a command timeout too (that will throw an exception if ```execute(String cmd)``` doesn't return within a given time. I'll update this issue once that is done as well.

**Comment by GideonLeGrange** [comment]
@prophe05 and @ruffboy 

Besides the optional connection timeout passed to ```connect()``` and ```connectTLS()``` I've added ```getTimeout()``` and ```setTimeout()``` methods to allow the user to adjust the timeouts for synchronous API commands. The connection timeout applies to the opening of the TCP socket only. So let's say for example you want commands to either return successfully within 2 seconds or fail with an exception. The you can do:

```java
ApiConnection con = ApiConnection.connect(HOSTNAME);
con.setTimeout(2000); // 2 seconds in milliseconds
con.login(USERNAME, PASSWORD);
// more commands from here
```

**Comment by prophe05** [comment]
Thx, it works for me)

**Comment by GideonLeGrange** [comment]
Changes discussed merged into master branch. Will release version 2.1 around the weekend.
